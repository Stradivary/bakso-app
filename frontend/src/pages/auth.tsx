import { Box, Button, Checkbox, Container, Flex, Select, Text, TextInput, Title } from '@mantine/core';
import React from 'react';
import { useNavigate } from 'react-router-dom';
import { useLocation } from '../shared/hooks/useLocation';
import { authService, useAuth } from '../shared/services/authService';

import { Controller, useForm } from 'react-hook-form';

export const AuthPage: React.FC = () => {
  const { control, handleSubmit, formState: { errors } } = useForm({
    defaultValues: {
      name: '',
      role: '',
    }
  });
  const navigate = useNavigate();
  const { location } = useLocation();
  const { login } = useAuth();

  const handleAuth = async (data: { name: string; role: string; }) => {
    if (data.name && data.role && location) {
      const userData = await authService.registerOrLogin(data.name, data.role, location);
      login(userData);

      localStorage.setItem('role', data.role);
      navigate('/');
    }
  };

  return (
    <Flex style={{ height: "100dvh", alignItems: "center" }}>
      <Container size="xs" style={{ marginBlock: 'auto', textAlign: 'center' }}>
        <Box mb="md">
          <img src="/video.png" alt="Illustration" style={{ width: 150, margin: '0 auto' }} />
        </Box>
        <Title order={2}>Verifikasi</Title>
        <Text mt="sm" mb="lg">
          Masukkan nama dan role Anda di bawah ini:
        </Text>

        <form onSubmit={handleSubmit(handleAuth)}>
          <Controller
            name="name"
            control={control}
            rules={{ required: 'Nama is required' }}
            render={({ field }) => (
              <TextInput
                label="Nama"
                placeholder="Masukkan nama"
                mb="sm"
                required
                {...field}
                error={errors.name?.message}
              />
            )}
          />
          <Controller
            name="role"
            control={control}
            rules={{ required: 'Role is required' }}
            render={({ field }) => (
              <Select
                label="Role"
                placeholder="Masukkan role"
                mb="md"
                required
                data={[
                  { value: 'customer', label: 'Customer' },
                  { value: 'seller', label: 'Seller' },
                ]}
                {...field}
                error={errors.role?.message}
              />
            )}
          />

          <Button fullWidth color="red" size="md" radius={56} type="submit">
            Join
          </Button>
        </form>

        <Checkbox
          mt="lg"
          label="Dengan menggunakan aplikasi ini Anda telah setuju untuk membagikan lokasi Anda kepada para tukang Bakso Keliling."
          defaultChecked
        />
      </Container>
    </Flex>
  );
};

export default AuthPage;
