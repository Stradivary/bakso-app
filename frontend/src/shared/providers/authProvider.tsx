import { PropsWithChildren, useState, useEffect } from "react";
import { User } from "../models/user";
import { AuthContext } from "../services/authService";


export function AuthProvider({ children }: Readonly<PropsWithChildren>) {
  const [user, setUser] = useState<User | null>(null);

  useEffect(() => {
    const storedUser = localStorage.getItem('user');
    if (storedUser) {
      setUser(JSON.parse(storedUser));
    }
  }, []);

  const login = (userData: User) => {
    setUser(userData);
    localStorage.setItem('user', JSON.stringify(userData));
  };

  const logout = () => {
    setUser(null);
    localStorage.removeItem('user');
  };

  return (
    <AuthContext.Provider value={{ user, login, logout }}>
      {children}
    </AuthContext.Provider>
  );
}
