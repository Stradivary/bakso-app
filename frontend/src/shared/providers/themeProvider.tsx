import { MantineProvider } from "@mantine/core";
import { ModalsProvider } from "@mantine/modals";
import { Notifications } from "@mantine/notifications";
import { PropsWithChildren } from "react";

export const ThemeProvider = ({ children }: PropsWithChildren) => {
  return (
    <MantineProvider theme={{ primaryColor: "red" }}>
      <ModalsProvider>
        {children}
        <Notifications position="top-center" />
      </ModalsProvider>
    </MantineProvider>
  );
};
