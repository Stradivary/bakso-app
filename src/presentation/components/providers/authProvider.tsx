import { Database } from "@/data/models/supabase.types";
import { signInUser } from "@/domain/services/authService";
import { supabase } from "@/domain/services/supabaseService";
import { notifications } from "@mantine/notifications";
import { Session } from "@supabase/supabase-js";
import {
  createContext,
  PropsWithChildren,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

type UserDetails = Database["public"]["Tables"]["user_profiles"]["Row"];

export interface AuthContextType {
  session: Session | null;
  user: UserDetails | null;
  isLoading: boolean;
  error: string | null;
  login: (
    name: string,
    role: string,
    latitude: number,
    longitude: number,
  ) => Promise<void>;
  logout: () => Promise<void>;
  isAuthenticated: boolean;
}

export const AuthProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [session, setSession] = useState<Session | null>(null);
  const [user, setUser] = useState<UserDetails | null>(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);
  const [initialized, setInitialized] = useState(false);

  // Fetch user profile data
  const fetchUserProfile = useCallback(async (userId: string) => {
    try {
      const { data, error: fetchError } = await supabase
        .from("user_profiles")
        .select("*")
        .eq("id", userId)
        .single();

      if (fetchError) throw fetchError;

      if (data) {
        setUser(data);
      }
    } catch (error) {
      console.error("Error fetching user profile:", error);
      setError("Failed to fetch user profile");
    }
  }, []);

  // Initialize auth state
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        // Check for existing session
        const {
          data: { session: existingSession },
          error: sessionError,
        } = await supabase.auth.getSession();

        if (sessionError) {
          console.error("failed to set session, log in again");
          throw sessionError;
        }

        if (existingSession) {
          setSession(existingSession);
          await fetchUserProfile(existingSession.user.id);
        }
      } catch (error) {
        console.error("Error initializing auth:", error);
        setError("Failed to initialize authentication");
      } finally {
        setIsLoading(false);
        setInitialized(true);
      }
    };

    initializeAuth();
  }, [fetchUserProfile]);
  const login = useCallback(
    async (name: string, role: string, latitude: number, longitude: number) => {
     
      try {
        setIsLoading(true);
        const { session: newSession, error: signInError } = await signInUser(
          name,
          role,
          latitude,
          longitude,
        );
        if (signInError) throw signInError;

        setSession(newSession);
        if (newSession) {
          await fetchUserProfile(newSession.user.id);
        }
        setError(null);
      } catch (error) {
        console.error("Error during login:", error);
        setError("Failed to login");
      } finally {
        setIsLoading(false);
      }
    },
    [fetchUserProfile],
  );

  const logout = useCallback(async () => {
    try {
      const { error: signOutError } = await supabase.auth.signOut({
        scope: "local",
      });

      if (signOutError) throw signOutError;
      setSession(null);
      setUser(null);
    } catch (error) {
      console.error("Error during logout:", error);
      setError("Failed to logout");
    }
  }, []);

  // Storage change observer
  useEffect(() => {
    if (!initialized) return;

    const handleStorageChange = (event: StorageEvent) => {
      const currentSession = event.newValue;
      const previousSession = event.oldValue;

      if (event.key === "abangbakso-session") {
        if (!previousSession) {
          notifications.show({
            title: "Warning",
            message: "Possible session injection detected",
            color: "orange",
          });
          logout();
          setError("Session security violation detected");
          return;
        }

        if (!currentSession) {
          notifications.show({
            title: "Warning",
            message: "Possible session removal detected",
            color: "orange",
          });
          logout();
          setError("Session security violation detected");
          return;
        }
        // If session was changed externally
        if (currentSession !== previousSession) {
          notifications.show({
            title: "Warning",
            message: "Possible session tampering detected",
            color: "orange",
          });
          logout();
          setError("Session security violation detected");
        }
      }
    };

    // Add storage event listener
    window.addEventListener("storage", handleStorageChange);

    return () => {
      window.removeEventListener("storage", handleStorageChange);
    };
  }, [initialized, logout, session?.user.id]);

  useEffect(() => {
    if (!initialized) return;

    const {
      data: { subscription },
    } = supabase.auth.onAuthStateChange(async (event, newSession) => { 

      if (!newSession?.user.id) return;
      if ((newSession?.user.id && newSession?.user.id) && newSession?.user.id !== session?.user.id) {
        return;
      }
      switch (event) {
        case "INITIAL_SESSION":
          break;
        case "SIGNED_IN":
        case "TOKEN_REFRESHED":
        case "USER_UPDATED":
          setSession(newSession);
          if (newSession) {
            await fetchUserProfile(newSession.user.id);
          }
          break;
        case "SIGNED_OUT":
          setSession(null);
          setUser(null);
          break;
      }
    });

    return () => {
      subscription.unsubscribe();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [initialized, fetchUserProfile]);

  const isAuthenticated = !!session?.user?.id && !isLoading;

  const values = useMemo(
    () => ({
      session,
      user,
      isLoading,
      error,
      login,
      logout,
      isAuthenticated,
    }),
    [session, user, isLoading, error, login, logout, isAuthenticated],
  );

  return <AuthContext.Provider value={values}>{children}</AuthContext.Provider>;
};

export const AuthContext = createContext<AuthContextType | undefined>(
  undefined,
);

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
