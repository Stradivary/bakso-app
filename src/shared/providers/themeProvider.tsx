import { createTheme, MantineColorsTuple, MantineProvider } from "@mantine/core";
import { ModalsProvider } from "@mantine/modals";
import { Notifications } from "@mantine/notifications";
import { PropsWithChildren } from "react";

const primary: MantineColorsTuple = [
  '#ffe8ed',
  '#ffcfd7',
  '#fe9cab',
  '#fd677d',
  '#fc3a56',
  '#fb203e',
  '#fc1131',
  '#e10224',
  '#c9001f',
  '#b00018'
];

const secondary: MantineColorsTuple = [
  '#ebf4ff',
  '#d4e5fa',
  '#a3c9f7',
  '#71abf6',
  '#4b92f5',
  '#3782f5',
  '#2e7af6',
  '#2368dc',
  '#195dc5',
  '#0050ad'
];

const theme = createTheme({

  colors: {
    primary,
    secondary,
  },
  primaryColor: 'primary',
  defaultRadius: 8,
  fontFamily: "'Poppins', system-ui, sans-serif",
  headings: {
    fontFamily: "'Telkomsel Batik Sans', system-ui, sans-serif",
  }
});

export const ThemeProvider = ({ children }: PropsWithChildren) => {
  return (
    <MantineProvider theme={theme}>
      <ModalsProvider>
        {children}
        <Notifications position="top-center" />
      </ModalsProvider>
    </MantineProvider>
  );
};
