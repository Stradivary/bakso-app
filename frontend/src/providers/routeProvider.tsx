import { createBrowserRouter, RouterProvider } from "react-router-dom";
import { ProtectedRoute } from "../components/ProtectedRoute";
import { Register, MainPage, CustomerPage } from "../pages";

const router = createBrowserRouter([
  {
    path: '/auth',
    element: <Register />,
  },
  {
    path: '/',
    element: (
      <ProtectedRoute>
        <CustomerPage />
      </ProtectedRoute>
    ),
  },
  {
    path: '/maps',
    element: <CustomerPage />,
  },

]);


export function RouteProvider() {
  return (
    <RouterProvider router={router}  />
  );
}