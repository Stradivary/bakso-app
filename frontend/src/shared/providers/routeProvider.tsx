import { createBrowserRouter, RouterProvider } from "react-router-dom";
import { ProtectedRoute } from "../components/ProtectedRoute";
import { AuthPage, CustomerPage } from "../../pages";

const router = createBrowserRouter([
  {
    path: '/auth',
    element: <AuthPage />,
  },
  {
    path: '/',
    element: (
      <ProtectedRoute>
        <CustomerPage />
      </ProtectedRoute>
    ),
  },
]);


export function RouteProvider() {
  return (
    <RouterProvider router={router} />
  );
}