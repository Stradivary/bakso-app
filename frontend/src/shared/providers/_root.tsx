import { MantineProvider } from "@mantine/core";
import { Notifications } from "@mantine/notifications";
import { AuthProvider } from "./authProvider";

import "@mantine/core/styles.css";
import "@mantine/notifications/styles.css";
import { RouteProvider } from "./routeProvider";


export function Root() {
  return (
    <AuthProvider>
      <MantineProvider theme={{ primaryColor: "red" }}>
        <RouteProvider />
        <Notifications />
      </MantineProvider>
    </AuthProvider>
  );
}

