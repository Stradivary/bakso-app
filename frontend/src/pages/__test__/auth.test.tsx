import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { beforeAll, describe, expect, it, vi } from "vitest";
import { ThemeProvider } from "../../shared/providers/themeProvider";
import { AuthPage } from "../auth";

// mock navigator.geolocation
const mockGeolocation = {
  getCurrentPosition: vi.fn(),
  watchPosition: vi.fn(),
};

// @ts-expect-error navigator is readonly
global.navigator.geolocation = mockGeolocation;

// mock ResizeObserver
class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}

beforeAll(() => {
  global.ResizeObserver = ResizeObserver;
});

describe("AuthPage", () => {
  it("should render correctly", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ThemeProvider>
          <AuthPage />
        </ThemeProvider>
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
