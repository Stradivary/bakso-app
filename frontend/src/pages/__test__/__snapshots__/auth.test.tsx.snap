// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthPage > should render correctly 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-primary-color-filled: var(--mantine-color-red-filled);--mantine-primary-color-filled-hover: var(--mantine-color-red-filled-hover);--mantine-primary-color-light: var(--mantine-color-red-light);--mantine-primary-color-light-hover: var(--mantine-color-red-light-hover);--mantine-primary-color-light-color: var(--mantine-color-red-light-color);--mantine-primary-color-0: var(--mantine-color-red-0);--mantine-primary-color-1: var(--mantine-color-red-1);--mantine-primary-color-2: var(--mantine-color-red-2);--mantine-primary-color-3: var(--mantine-color-red-3);--mantine-primary-color-4: var(--mantine-color-red-4);--mantine-primary-color-5: var(--mantine-color-red-5);--mantine-primary-color-6: var(--mantine-color-red-6);--mantine-primary-color-7: var(--mantine-color-red-7);--mantine-primary-color-8: var(--mantine-color-red-8);--mantine-primary-color-9: var(--mantine-color-red-9);}:root[data-mantine-color-scheme="dark"]{--mantine-color-anchor: var(--mantine-color-red-4);}:root[data-mantine-color-scheme="light"]{--mantine-color-anchor: var(--mantine-color-red-6);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m_8bffd616 mantine-Flex-root __m__-r1"
    style="align-items: center;"
  >
    <div
      class="m_7485cace mantine-Container-root"
      data-size="xs"
      style="--container-size: var(--container-size-xs); margin-block: auto; text-align: center;"
    >
      <div
        class=""
      >
        <img
          alt="Illustration"
          src="/video.png"
          style="width: 150px; margin: 0px auto;"
        />
      </div>
      <h2
        class="m_8a5d1357 mantine-Title-root"
        data-order="2"
        style="--title-fw: var(--mantine-h2-font-weight); --title-lh: var(--mantine-h2-line-height); --title-fz: var(--mantine-h2-font-size);"
      >
        Verifikasi
      </h2>
      <p
        class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
      >
        Masukkan nama dan role Anda di bawah ini:
      </p>
      <form
        style="text-align: start;"
      >
        <div
          class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
        >
          <label
            class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
            for=""
            id=""
          >
            Nama
          </label>
          <div
            class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
            data-variant="default"
          >
            <input
              aria-invalid="false"
              class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
              data-variant="default"
              id=""
              name="name"
              placeholder="Masukkan nama"
              value=""
            />
          </div>
        </div>
        <div
          class="m_46b77525 mantine-InputWrapper-root mantine-Select-root"
        >
          <label
            class="m_8fdc1311 mantine-InputWrapper-label mantine-Select-label"
            for=""
            id=""
          >
            Role
          </label>
          <div
            class="m_6c018570 mantine-Input-wrapper mantine-Select-wrapper"
            data-pointer="true"
            data-variant="default"
            data-with-right-section="true"
            style="--input-right-section-pointer-events: none;"
          >
            <input
              aria-haspopup="listbox"
              aria-invalid="false"
              autocomplete="off"
              class="m_8fb7ebe7 mantine-Input-input mantine-Select-input"
              data-variant="default"
              id=""
              placeholder="Masukkan role"
              readonly=""
              value=""
            />
            <div
              class="m_82577fc2 mantine-Input-section mantine-Select-section"
              data-position="right"
            >
              <svg
                class="m_2943220b mantine-ComboboxChevron-chevron"
                data-combobox-chevron="true"
                fill="none"
                viewBox="0 0 15 15"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="role"
          type="hidden"
          value=""
        />
        <button
          class="mantine-focus-auto m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
          data-block="true"
          data-disabled="true"
          data-size="md"
          disabled=""
          style="--button-height: var(--button-height-md); --button-padding-x: var(--button-padding-x-md); --button-fz: var(--mantine-font-size-md); --button-radius: calc(3.5rem * var(--mantine-scale)); --button-bg: var(--mantine-color-red-filled); --button-hover: var(--mantine-color-red-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
          type="submit"
        >
          <span
            class="m_80f1301b mantine-Button-inner"
          >
            <span
              class="m_811560b9 mantine-Button-label"
            >
              Join
            </span>
          </span>
        </button>
        <div
          class="m_bf2d988c mantine-Checkbox-root m_5f75b09e mantine-Checkbox-root"
          data-label-position="right"
          style="--checkbox-color: var(--mantine-color-red-filled);"
        >
          <div
            class="m_5f6e695e mantine-Checkbox-body"
          >
            <div
              class="m_26062bec mantine-Checkbox-inner"
              data-label-position="right"
            >
              <input
                class="mantine-focus-auto m_26063560 mantine-Checkbox-input"
                id=""
                name="agreed"
                type="checkbox"
              />
              <svg
                aria-hidden="true"
                class="m_bf295423 mantine-Checkbox-icon"
                fill="none"
                viewBox="0 0 10 7"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M4 4.586L1.707 2.293A1 1 0 1 0 .293 3.707l3 3a.997.997 0 0 0 1.414 0l5-5A1 1 0 1 0 8.293.293L4 4.586z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
            <div
              class="m_d3ea56bb mantine-Checkbox-labelWrapper"
            >
              <label
                class="m_8ee546b8 mantine-Checkbox-label"
                for=""
              >
                Dengan menggunakan aplikasi ini Anda telah setuju untuk membagikan lokasi Anda kepada para tukang Bakso Keliling.
              </label>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
